### PR TITLE
Add Rubocop file for consistent styling

### PR DESCRIPTION
**Pull request by hectron**
# What

This pull request adds a **.rubocop.yml** file to the repository, so that we can have consistent styling. Additionally, it only runs Rubocop against a pull request since these syntax errors should be caught at the PR stage.

## Misc

It'd be nice to start updating the rest of the `happy-software` Ruby repos to use the same Rubocop settings. We can even
have a main repo that has our preferred styles/configurations, which we can all contribute to.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -13,6 +13,7 @@ jobs:
   rubocop-lint:
     name: Rubocop
     runs-on: ubuntu-latest
+    if: github.event_name == 'pull_request'
```
Comment (hectron) [inline .github/workflows/ci.yml:16]: This is to run Rubocop only on pull-requests. This is useful because we won't actually catch Rubocop errors on merge unless something really terrible went wrong.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -24,9 +25,12 @@ jobs:
       run: gem install rubocop
 
     - name: Run Rubocop
-      run: rubocop
       # Allow Rubocop to run without breaking the workflow
       continue-on-error: true
+      # Only run Rubocop on changed files
+      run: |
+        git fetch origin ${{ github.event.pull_request.base.sha }}
```
Comment (hectron) [inline .github/workflows/ci.yml:32]: This little snippet was required because the `actions/checkout@v2` action does a **shallow** clone of the repo -- it only checks out the relevant SHA so that the CI/CD pipeline goes faster. 

Without this when the repo is cloned in CI/CD, it does not have the HEAD commit on master; without having the HEAD commit on master, we can't deduce what files changed in this PR.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -24,9 +25,12 @@ jobs:
...
+        git diff -r --name-only ${{ github.event.pull_request.base.sha }} | grep '\.rb$' | xargs -r rubocop
```
Comment (hectron) [inline .github/workflows/ci.yml:33]: This returns the files that were changed on this pull request, and we only double check Ruby files. If there are any files, we run Rubocop against them.

File: .rubocop.yml [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+# We don't need screaming snake case all the damn time
+Naming/ConstantName:
+  Enabled: false
+
+# Single quotes being faster is hardly measurable and only affects parse time.
+# Enforcing double quotes reduces the times where you need to change them
+# when introducing an interpolation. Use single quotes only if their semantics
+# are needed.
+Style/StringLiterals:
+  EnforcedStyle: double_quotes
```
Comment (hebron-george) [inline .rubocop.yml:42]: I'm a fan of single quotes, not gonna lie.
We use this double quote rule in some other repos I've used at work and don't see a benefit personally.
Should we enforce this rule?

File: .rubocop.yml [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+Style/StringLiteralsInInterpolation:
+  EnforcedStyle: double_quotes
+
+Style/FrozenStringLiteralComment:
+  Enabled: false
```
Comment (hebron-george) [inline .rubocop.yml:48]: This might be nice because it makes all string literals frozen and reduces memory usage

File: .rubocop.yml [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+Style/Documentation:
+  Enabled: false
+
+Style/TrailingCommaInArguments:
+  EnforcedStyleForMultiline: consistent_comma
+
+Style/TrailingCommaInHashLiteral:
+  EnforcedStyleForMultiline: consistent_comma
```
Comment (hebron-george) [inline .rubocop.yml:57]: Does is cop mean that there is or isn't a trailing comma?

File: .github/workflows/ci.yml (hunk 2 above)
Comment (hebron-george) [reply .github/workflows/ci.yml:32]: So now we do the shallow clone + a fetch of the PR's base (i.e. `master`) right?

Comment (hectron) [reply .github/workflows/ci.yml:32]: That's right!

File: .rubocop.yml (hunk 4 above)
Comment (hectron) [reply .rubocop.yml:42]: I included this because it was consistent at Braintree and also highlighted in the Shopify Ruby style guide: https://github.com/Shopify/ruby-style-guide/blob/main/rubocop.yml

I personally prefer having a single type of quotes, so I'm open to either or.

File: .rubocop.yml (hunk 6 above)
Comment (hectron) [reply .rubocop.yml:57]: I think this means there is a trailing comma if it's a multi-line hash:

```ruby
# bad
a = { foo: 1, bar: 2, }

# good
a = { foo: 1, bar: 2 }

# good
a = {
  foo: 1, bar: 2,
  qux: 3,
}

# good
a = {
  foo: 1, bar: 2, qux: 3,
}

# good
a = {
  foo: 1,
  bar: 2,
}
```

File: .rubocop.yml (hunk 5 above)
Comment (hectron) [reply .rubocop.yml]: Agreed! I've updated this rule to be auto-correctable. For more details, check out the description of this commit:

90728ca3d2b13e4aad709fe265af26b4cc80bfdf

TL;DR -- we can run the following to make sure it's enforced across the repo:

```zsh
rubocop --auto-correct
```

File: .rubocop.yml (hunk 4 above)
Comment (hebron-george) [reply .rubocop.yml:42]: I say let's not enforce quote rules 🤷🏽

File: .rubocop.yml [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    - spec/**/*
+
+# Too short methods lead to extraction of single-use methods, which can make
+# the code easier to read (by naming things), but can also clutter the class
+Metrics/MethodLength: 
+  Max: 20
+
+# The guiding principle of classes is SRP, SRP can't be accurately measured by LoC
+Metrics/ClassLength:
+  Max: 1500
```
Comment (hebron-george) [inline .rubocop.yml:31]: 1500 seems very large for a single file.
I'm open to cutting it down by a lot if you want.
500?

File: .rubocop.yml (hunk 4 above)
Comment (hectron) [reply .rubocop.yml]: Updated to remove checking single vs double quotes

File: .rubocop.yml (hunk 7 above)
Comment (hectron) [reply .rubocop.yml]: Word! Updated to 500.

**Review by hebron-george — APPROVED** [review]
